src/components/drawing/index.js

```diff
@@ -355,88 +425,36 @@ function singlePointStyle(d, sel, trace, markerScale, lineScale, marker, markerL
             if(perPointGradient) gradientID += '-' + d.i;
 
             sel.call(drawing.gradient, gd, gradientID, gradientType, fillColor, gradientColor);
-        }
-        else {
+        } else {
             sel.call(Color.fill, fillColor);
         }
 
         if(lineWidth) {
             sel.call(Color.stroke, lineColor);
         }
     }
-}
-
-var HORZGRADIENT = {x1: 1, x2: 0, y1: 0, y2: 0};
-var VERTGRADIENT = {x1: 0, x2: 0, y1: 1, y2: 0};
-
-drawing.gradient = function(sel, gd, gradientID, type, color1, color2) {
-    var gradient = gd._fullLayout._defs.select('.gradients')
-        .selectAll('#' + gradientID)
-        .data([type + color1 + color2], Lib.identity);
-
-    gradient.exit().remove();
-
-    gradient.enter()
-        .append(type === 'radial' ? 'radialGradient' : 'linearGradient')
-        .each(function() {
-            var el = d3.select(this);
-            if(type === 'horizontal') el.attr(HORZGRADIENT);
-            else if(type === 'vertical') el.attr(VERTGRADIENT);
-
-            el.attr('id', gradientID);
-
-            var tc1 = tinycolor(color1);
-            var tc2 = tinycolor(color2);
-
-            el.append('stop').attr({
-                offset: '0%',
-                'stop-color': Color.tinyRGB(tc2),
-                'stop-opacity': tc2.getAlpha()
-            });
-
-            el.append('stop').attr({
-                offset: '100%',
-                'stop-color': Color.tinyRGB(tc1),
-                'stop-opacity': tc1.getAlpha()
-            });
-        });
-
-    sel.style({
-        fill: 'url(#' + gradientID + ')',
-        'fill-opacity': null
-    });
-};
-
-/*
- * Make the gradients container and clear out any previous gradients.
- * We never collect all the gradients we need in one place,
- * so we can't ever remove gradients that have stopped being useful,
- * except all at once before a full redraw.
- * The upside of this is arbitrary points can share gradient defs
- */
-drawing.initGradients = function(gd) {
-    var gradientsGroup = Lib.ensureSingle(gd._fullLayout._defs, 'g', 'gradients');
-    gradientsGroup.selectAll('linearGradient,radialGradient').remove();
 };
 
-drawing.singlePointStyle = function(d, sel, trace, markerScale, lineScale, gd) {
+drawing.makePointStyleFns = function(trace) {
```

Comment:
Define trace-wide per-pt functions w/o knowledge of d3 or DOM method so that non-svg module (e.g. `scattermapbox`) can reuse them :recycle: